<a href="https://colab.research.google.com/github/trupsihonyashka/lab-git/blob/main/homework/lesson-02/solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Урок 2. Основы Python: переменные, типы и условия

**Имя и фамилия: Мария Хисиева**
**Группа: 25-ИИбо-1а**

Это единый Notebook для практики на занятии и домашней работы. Работайте в
нём во время урока, сохраните результат, а после выполнения домашней части
загрузите тот же файл как `homework/lesson-02/solution.ipynb`.

## 1. Проверяем рабочий цикл

Выполните ячейку. Под числом сформулируйте словами, что оно означает.

In [43]:
1250 * 3

3750

**Мой вывод:** После выполнения ячейки вывело результат ее работы: произведение двух чисел ячейки: 1250 * 3 = 3750

## 2. Переменные и состояние Kernel

`=` присваивает значение имени. Для нескольких слов используем `snake_case`.
Python определяет тип значения во время выполнения.

In [44]:
product_name = "Наушники"
price = 1250
quantity = 3
order_total = price * quantity

print(product_name, order_total)

Наушники 3750


### Практика 1. Изменяем заказ

1. До запуска спрогнозируйте результат.
2. Создайте свои `price` и `quantity`.
3. Рассчитайте `order_total`.
4. Измените количество и выполните ячейку повторно.
5. Исправьте имя `order_totall`, не запуская неверный вариант повторно.

In [45]:
# Напишите решение здесь.
price = 1488
quantity = 67
order_total = price * quantity
print(order_total)

99696


**Что изменилось и почему:** При изменении количества и повторном запуске ячейки order_total изменится (т.к. пересичтается с новыми значениями). Если order_total изменить в одном месте на order_totall, то появится ошибка (NameError (интерпретатор не найдет переменную с таким именем в памяти.)), если изменить имя в обоих местах, то ячейка упсешно выоплнится

## 3. Типы и преобразования

`int` — целое число, `float` — дробное, `str` — текст, `bool` — логическое
значение, `None` — значение пока отсутствует. `type()` проверяет тип.

In [46]:
product_name = "Клавиатура"
price = 3200
discount_rate = 0.05
is_member = True
promo_code = None

type(product_name), type(price), type(discount_rate), type(is_member), type(promo_code)

(str, int, float, bool, NoneType)

In [47]:
quantity_text = "2"
quantity = int(quantity_text)
order_total = price * quantity

print(order_total, type(quantity_text), type(quantity))

6400 <class 'str'> <class 'int'>


### Практика 2. Исправляем тип

Код `"Цена: " + 3200` вызывает ошибку. Исправьте его двумя способами:
через `str()` и через f-строку. Объясните причину ошибки.

In [48]:
price = 3200
# Напишите два исправленных варианта.

In [49]:
"Цена: " + str(price)

'Цена: 3200'

In [50]:
(f"Цена: {price}")

'Цена: 3200'

In [51]:
"Цена: 3200"

'Цена: 3200'

**Причина ошибки:** Попытка сложить разные типы данных

## 4. Строки и f-строки

f-строка вставляет значения и результаты выражений внутрь понятного текста.

In [52]:
product_name = "Клавиатура"
price = 3200
quantity = 2
order_total = price * quantity

print(f"{quantity} × {product_name}: {order_total} рублей")

2 × Клавиатура: 6400 рублей


### Практика 3. Понятное сообщение

Выведите одной f-строкой название, цену, количество, сумму и наличие
клубной карты. Измените данные и обновите результат.

In [53]:
# Напишите решение здесь.
product_name = "попрыгунчик"
price = 1488
quantity = 67
order_total = price * quantity
member = True

message = (f"Товар: {product_name}, Цена: {price}, Количество: {quantity}, Сумма: {order_total}, Наличие клубной карты: {member}")

message

'Товар: попрыгунчик, Цена: 1488, Количество: 67, Сумма: 99696, Наличие клубной карты: True'

## 5. Сравнения и логические операции

`= ` присваивает, `==` сравнивает. Операторы `!=`, `>`, `<`, `>=`, `<=`
возвращают `True` или `False`. Условия объединяются через `and`, `or`, `not`.

In [54]:
free_delivery_threshold = 5000

print(4999 >= free_delivery_threshold)
print(5000 >= free_delivery_threshold)
print(5001 >= free_delivery_threshold)

False
True
True


### Практика 4. Предсказываем результат

Проверьте правило: скидка доступна, если сумма не меньше 5000 **и** есть
клубная карта. Используйте три заданных набора:

1. `order_total = 4999`, `is_member = True`;
2. `order_total = 5000`, `is_member = False`;
3. `order_total = 5000`, `is_member = True`.

Перед запуском запишите прогноз `True` или `False`, затем проверьте его.

Исправьте ошибку `if order_total = 5000:` и объясните замену оператора.

In [55]:
# Выполните проверки здесь.
sum = 5000
is_member = True

In [56]:
order_total = 4999
is_member = True
print(order_total >= sum and is_member == True)

False


In [57]:
order_total = 5000
is_member = False
print(order_total >= sum and is_member == True)

False


In [58]:
order_total = 5000
is_member = True
print(order_total >= sum and is_member == True)

True


**Мой вывод о границе и составном условии:** False, False, True, В услвоии использован неверный оператор. Вместо оператора сравнения (==) использован оператор присвоения (=). Исправленный вариант: if order_total == 5000 / if order_total >= 5000 / if order_total <= 5000 / if order_total != 5000

## 6. `if` / `elif` / `else`

Python проверяет ветви сверху вниз и выполняет первую подходящую. После
условия ставится двоеточие, команды внутри блока имеют одинаковый отступ.

In [59]:
order_total = 6500

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 5000:
    discount_rate = 0.05
else:
    discount_rate = 0

total_after_discount = order_total * (1 - discount_rate)
print(f"Скидка: {discount_rate:.0%}; после скидки: {total_after_discount:.2f} руб.")

Скидка: 5%; после скидки: 6175.00 руб.


### Практика 5. Исправляем ошибки

Перенесите фрагмент в рабочую ячейку и исправьте двоеточие, отступы и
недостающую ветвь. Затем проверьте 4999, 5000, 6999 и 7000.

```python
if order_total >= 7000
discount_rate = 0.10
elif order_total >= 5000:
    discount_rate = 0.05
```

In [60]:
# Исправленный код и проверки.

if order_total >= 7000:
  discount_rate = 0.10
elif order_total >= 5000:
  discount_rate = 0.05
else:
  discount_rate = 0

# Проверка 4999

order_total = 4999

if order_total >= 7000:
  discount_rate = 0.10
elif order_total >= 5000:
  discount_rate = 0.05
else:
  discount_rate = 0
print(discount_rate)

# Проверка 5000

order_total = 5000

if order_total >= 7000:
  discount_rate = 0.10
elif order_total >= 5000:
  discount_rate = 0.05
else:
  discount_rate = 0
print(discount_rate)

# Проверка 6999

order_total = 6999

if order_total >= 7000:
  discount_rate = 0.10
elif order_total >= 5000:
  discount_rate = 0.05
else:
  discount_rate = 0
print(discount_rate)

# Проверка 7000

order_total = 7000

if order_total >= 7000:
  discount_rate = 0.10
elif order_total >= 5000:
  discount_rate = 0.05
else:
  discount_rate = 0
print(discount_rate)

0
0.05
0.05
0.1


**Какие ошибки были исправлены:** Было добавлено добавлено ":" после `if order_total >= 7000`, добавлен отступ перед `discount_rate = 0.10`. Была доабвлена недостающая ветвь:
`
else:
  discount_rate = 0
`




## 7. Самостоятельная практика «Итог заказа»

**Исходные правила:** скидка 10% при сумме от 7000; скидка 5% при сумме
от 5000 и наличии клубной карты; иначе скидка 0%. Доставка бесплатна, если
сумма после скидки не меньше 5000, иначе доставка стоит 300 рублей.

**Что сделать:**

1. Создайте `product_name`, `price`, `quantity`, `is_member`.
2. Рассчитайте `order_total`.
3. Через `if` / `elif` / `else` присвойте `discount_rate`.
4. Рассчитайте `total_after_discount`.
5. Через второе условие присвойте `actual_delivery_cost`.
6. Рассчитайте `final_total` и выведите все результаты одной f-строкой.
7. Проверьте три сценария: без скидки; со скидкой 5%; ровно 7000 до скидки.
8. Для каждого сценария заполните подготовленное поле ниже.

**Готово, если:** код выполняется без ошибок, значения меняются внутри
ветвей, проверены три сценария и вывод соответствует результату.

In [61]:
#Сценарий 1: без скидки

# 1. Исходные данные

product_name = "Попрыгунчик"
price = 1000
quantity = 2
is_member = False

# 2. Исходная сумма

order_total = price * quantity

# 3. Выбор discount_rate

if (order_total >= 7000):
  discount_rate = 0.1
elif((order_total >= 5000) and (is_member == True)):
  discount_rate = 0.05
else:
  discount_rate = 0

# 4. Сумма после скидки

total_after_discount = order_total * (1 - discount_rate)

# 5. Выбор actual_delivery_cost

if(total_after_discount >= 5000):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = 300

# 6. Итоговая сумма и f-строка

final_total = total_after_discount + actual_delivery_cost
print(f"Товар: {product_name};\n"
      f"Количество: {quantity};\n"
      f"Цена за шт.: {price};\n"
      f"Наличие клубной карты: {is_member};\n"
      f"Сумма покупки без учета скидки: {order_total};\n"
      f"Скидка: {discount_rate};\n"
      f"Сумма покупки с учетом скидки: {total_after_discount};\n"
      f"Стоимость доставки: {actual_delivery_cost};\n"
      f"Сумма покупки с учетом скидки и стоимости доставки: {final_total};")

Товар: Попрыгунчик;
Количество: 2;
Цена за шт.: 1000;
Наличие клубной карты: False;
Сумма покупки без учета скидки: 2000;
Скидка: 0;
Сумма покупки с учетом скидки: 2000;
Стоимость доставки: 300;
Сумма покупки с учетом скидки и стоимости доставки: 2300;


In [62]:
#Сценарий 2: со скидкой 5%

# 1. Исходные данные

product_name = "Попрыгунчик"
price = 100
quantity = 67
is_member = True

# 2. Исходная сумма

order_total = price * quantity

# 3. Выбор discount_rate

if (order_total >= 7000):
  discount_rate = 0.1
elif((order_total >= 5000) and (is_member == True)):
  discount_rate = 0.05
else:
  discount_rate = 0

# 4. Сумма после скидки

total_after_discount = order_total * (1 - discount_rate)

# 5. Выбор actual_delivery_cost

if(total_after_discount >= 5000):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = 300

# 6. Итоговая сумма и f-строка

final_total = total_after_discount + actual_delivery_cost
print(f"Товар: {product_name};\n"
      f"Количество: {quantity};\n"
      f"Цена за шт.: {price};\n"
      f"Наличие клубной карты: {is_member};\n"
      f"Сумма покупки без учета скидки: {order_total};\n"
      f"Скидка: {discount_rate};\n"
      f"Сумма покупки с учетом скидки: {total_after_discount};\n"
      f"Стоимость доставки: {actual_delivery_cost};\n"
      f"Сумма покупки с учетом скидки и стоимости доставки: {final_total};")

Товар: Попрыгунчик;
Количество: 67;
Цена за шт.: 100;
Наличие клубной карты: True;
Сумма покупки без учета скидки: 6700;
Скидка: 0.05;
Сумма покупки с учетом скидки: 6365.0;
Стоимость доставки: 0;
Сумма покупки с учетом скидки и стоимости доставки: 6365.0;


In [63]:
#Сценарий 3: 7000 до скидки

# 1. Исходные данные

product_name = "Попрыгунчик"
price = 700
quantity = 10
is_member = True

# 2. Исходная сумма

order_total = price * quantity

# 3. Выбор discount_rate

if (order_total >= 7000):
  discount_rate = 0.1
elif((order_total >= 5000) and (is_member == True)):
  discount_rate = 0.05
else:
  discount_rate = 0

# 4. Сумма после скидки

total_after_discount = order_total * (1 - discount_rate)

# 5. Выбор actual_delivery_cost

if(total_after_discount >= 5000):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = 300

# 6. Итоговая сумма и f-строка

final_total = total_after_discount + actual_delivery_cost
print(f"Товар: {product_name};\n"
      f"Количество: {quantity};\n"
      f"Цена за шт.: {price};\n"
      f"Наличие клубной карты: {is_member};\n"
      f"Сумма покупки без учета скидки: {order_total};\n"
      f"Скидка: {discount_rate};\n"
      f"Сумма покупки с учетом скидки: {total_after_discount};\n"
      f"Стоимость доставки: {actual_delivery_cost};\n"
      f"Сумма покупки с учетом скидки и стоимости доставки: {final_total};")

Товар: Попрыгунчик;
Количество: 10;
Цена за шт.: 700;
Наличие клубной карты: True;
Сумма покупки без учета скидки: 7000;
Скидка: 0.1;
Сумма покупки с учетом скидки: 6300.0;
Стоимость доставки: 0;
Сумма покупки с учетом скидки и стоимости доставки: 6300.0;


### Результаты проверок

**Сценарий 1:**  Сумма покупки без учета скидки: 2000; Скидка: 0; Сумма покупки с учетом скидки:  2000; Стоимость доставки: 300; Сумма покупки с учетом скидки и стоимости доставки:  2300;
**Результат и вывод:** Сумма покупки без учета скидки: 2000; Скидка: 0; Сумма покупки с учетом скидки:  2000; Стоимость доставки: 300; Сумма покупки с учетом скидки и стоимости доставки:  2300;

**Сценарий 2:** Наличие клубной карта: True; Сумма покупки без учета скидки: 6700; Скидка: 0.05%; Сумма покупки с учетом скидки:  6365.0; Стоимость доставки: 0; Сумма покупки с учетом скидки и стоимости доставки:  6365.0;
**Результат и вывод:** Наличие клубной карта: True; Сумма покупки без учета скидки: 6700; Скидка: 0.05%; Сумма покупки с учетом скидки:  6365.0; Стоимость доставки: 0; Сумма покупки с учетом скидки и стоимости доставки:  6365.0;

**Сценарий 3 — граничное значение:** Сумма покупки без учета скидки: 7000; Скидка: 0.1%; Сумма покупки с учетом скидки:  6300.0; Стоимость доставки: 0; Сумма покупки с учетом скидки и стоимости доставки:  6300.0;  
**Результат и вывод:** Сумма покупки без учета скидки: 7000; Скидка: 0.1%; Сумма покупки с учетом скидки:  6300.0; Стоимость доставки: 0; Сумма покупки с учетом скидки и стоимости доставки:  6300.0;

## 8. Кратко: `input()` и `while`

`input()` всегда возвращает строку. Для числа требуется преобразование:

```python
quantity = int(input("Количество: "))
```

Этот пример не запускаем через `Run All`, чтобы Notebook не ожидал ввод.
`while` повторяет блок, пока условие истинно; условие должно изменяться.

In [64]:
attempt = 0
max_attempts = 3

while attempt < max_attempts:
    print(f"Попытка {attempt + 1}")
    attempt += 1

Попытка 1
Попытка 2
Попытка 3


### Найдите риск бесконечного цикла

Почему код ниже не завершится? Что нужно изменить?

```python
attempt = 0
while attempt < 3:
    print(attempt)
```

**Мой ответ:**  Потому что attempt всегда будет < 3. Т.к. она никак не изменяется внутри цикла.
Одно из возможных испарвлений:
```
attempt = 0
while attempt < 3:
  print(attempt)
  attempt+=1

```



# Домашняя работа

Продолжайте работать в этом же Notebook. После выполнения загрузите его как
`homework/lesson-02/solution.ipynb`.

## Мини-исследование «Какие правила заказа выгодны покупателю?»

Все задания выполняются **ниже в уже подготовленных ячейках**. Новые ячейки
для обязательной части создавать не требуется.

**Обязательный результат:** базовый расчёт и четыре заданных эксперимента.
Для каждого эксперимента сохраните код, числовой результат и вывод словами.
Затем выполните минимум два задания расширенного уровня.

## Вопрос и гипотеза

**Мой вопрос:** По итогам напсианяи и выполненяи программы будет ли применена скидка и какова будет стоимость доставки?
**Моя гипотеза:**  Я предполагаю, что при цене товара 1400 руб. и количестве 3 шт. начальная сумма заказа составит 4200 руб. Так как эта сумма меньше порога для скидки (5000 руб.), даже несмотря на наличие клубной карты и меньше порога для бесплатной доставки (6000 руб.), скидка не будет применена (0%), а к сумме добавится стоимость доставки (400 руб.). Итоговая сумма (final_total) должна составить 4600 рублей.

In [65]:
# Задание 1. Базовая модель заказа
product_name = "Наушники"
price = 1400
quantity = 3
is_member = True
discount_threshold = 5000
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте order_total.
order_total = price * quantity

# Через if / else присвойте applied_discount_rate.
if(order_total >= discount_threshold and is_member == True):
  applied_discount_rate = discount_rate
else:
  applied_discount_rate = 0

# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - applied_discount_rate)

# Через второе if / else присвойте actual_delivery_cost.
if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

# Рассчитайте final_total и выведите понятную f-строку.
final_total = total_after_discount + actual_delivery_cost

print(f"Товар: {product_name}\n"
      f"Количество: {quantity}\n"
      f"Цена за шт.: {price}\n"
      f"Наличие клубной карты: {is_member}\n"
      f"Сумма покупки без учета скидки: {order_total}\n"
      f"Скидка: {applied_discount_rate}\n"
      f"Сумма покупки с учетом скидки: {total_after_discount}\n"
      f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")

Товар: Наушники
Количество: 3
Цена за шт.: 1400
Наличие клубной карты: True
Сумма покупки без учета скидки: 4200
Скидка: 0
Сумма покупки с учетом скидки: 4200
Стоимость доставки: 400
ИТОГО к оплате: 4600


## Обязательные эксперименты

В каждом блоке используйте правила базовой модели. Меняйте только указанные
данные, выполняйте расчёт полностью и заполняйте вывод под Code-ячейкой.

In [66]:
# Эксперимент 1: ниже порога скидки
# Используйте price = 1200, quantity = 4.
# Рассчитайте скидку, доставку и final_total.

product_name = "Наушники"
price = 1200
quantity = 4
is_member = True
discount_threshold = 5000
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте order_total.
order_total = price * quantity

# Через if / else присвойте applied_discount_rate.
if(order_total >= discount_threshold and is_member == True):
  applied_discount_rate = discount_rate
else:
  applied_discount_rate = 0

# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - applied_discount_rate)

# Через второе if / else присвойте actual_delivery_cost.
if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

# Рассчитайте final_total и выведите понятную f-строку.
final_total = total_after_discount + actual_delivery_cost

print(f"Скидка: {applied_discount_rate}\n"
      f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")


Скидка: 0
Стоимость доставки: 400
ИТОГО к оплате: 5200


**Вывод 1:** Скидка не применилась, так как не был достигнут порог для её получения (4800 < 5000). Доставка была платной по той же причине: сумма заказа не достигла порога бесплатной доставки (4800 < 6000). Итоговая сумма к оплате составила 5200.

In [67]:
# Эксперимент 2: ровно на пороге скидки
# Используйте price = 1000, quantity = 5.
# Проверьте, применяется ли скидка при order_total == discount_threshold.
product_name = "Наушники"
price = 1000
quantity = 5
is_member = True
discount_threshold = 5000
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте order_total.
order_total = price * quantity

# Через if / else присвойте applied_discount_rate.
if(order_total >= discount_threshold and is_member == True):
  applied_discount_rate = discount_rate
else:
  applied_discount_rate = 0

# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - applied_discount_rate)

# Через второе if / else присвойте actual_delivery_cost.
if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

# Рассчитайте final_total и выведите понятную f-строку.
final_total = total_after_discount + actual_delivery_cost

print(f"Скидка: {applied_discount_rate}\n"
      f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")


Скидка: 0.1
Стоимость доставки: 400
ИТОГО к оплате: 4900.0


**Вывод 2:** Скидка применилась благодаря оператору >= (больше или равно), так как сумма заказа (5000) равна порогу. Если бы использовался оператор > (строго больше), скидка бы не сработала, поскольку 5000 не больше 5000.
После применения скидки 10% сумма составила 4500. Так как это всё ещё ниже порога бесплатной доставки (6000), доставка осталась платной (400). Итоговая сумма к оплате: 4900.


In [68]:
# Эксперимент 3: граница бесплатной доставки
# Последовательно проверьте total_after_discount: 5999, 6000 и 6001.
# Для каждого значения определите actual_delivery_cost и final_total.
free_delivery_threshold = 6000
total_after_discount = 5999

if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

final_total = total_after_discount + actual_delivery_cost

print(f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")

total_after_discount = 6000

if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

final_total = total_after_discount + actual_delivery_cost

print(f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")

total_after_discount = 6001

if(total_after_discount >= free_delivery_threshold):
  actual_delivery_cost = 0
else:
  actual_delivery_cost = delivery_cost

final_total = total_after_discount + actual_delivery_cost

print(f"Стоимость доставки: {actual_delivery_cost}\n"
      f"ИТОГО к оплате: {final_total}")


Стоимость доставки: 400
ИТОГО к оплате: 6399
Стоимость доставки: 0
ИТОГО к оплате: 6000
Стоимость доставки: 0
ИТОГО к оплате: 6001


**Вывод 3:**  Бесплатная доставка применилась для значений total_after_discount = 6000 и 6001, так как оба этих значения больше или равны порогу free_delivery_threshold = 6000. Использование оператора >= гарантирует, что на самой границе условие считается выполненным. В то же время, при сумме 5999 доставка осталась платной, так как это значение строго меньше порогового, и правило не сработало.

## Эксперимент 4. Конфликт правил доставки

Используйте `order_total = 6500`, скидку 10% и порог бесплатной доставки
6000 рублей. Сравните два правила:

1. право на бесплатную доставку определяется **до скидки**;
2. право на бесплатную доставку определяется **после скидки**.

Рассчитайте две итоговые суммы.

In [69]:
order_total = 6500
discount_rate = 0.10
free_delivery_threshold = 6000
delivery_cost = 400

# Рассчитайте total_after_discount.
total_after_discount = order_total * (1 - discount_rate)

# Правило 1: определите delivery_before_discount и итоговую сумму.
if(order_total >= free_delivery_threshold):
  delivery_before_discount = 0
else:
  delivery_before_discount = delivery_cost
final_total = total_after_discount + delivery_before_discount
print(f"ИТОГО к оплате: {final_total}")

# Правило 2: определите delivery_after_discount и итоговую сумму.
if(total_after_discount >= free_delivery_threshold):
  delivery_before_discount = 0
else:
  delivery_before_discount = delivery_cost
final_total = total_after_discount + delivery_before_discount
print(f"ИТОГО к оплате: {final_total}")

ИТОГО к оплате: 5850.0
ИТОГО к оплате: 6250.0


**Вывод 4:** Покупателю выгоднее первое правило, так как итоговая сумма заказа ниже (5850 ₽ вместо 6250 ₽). Магазину выгоднее второе правило, так как покупатель оплачивает товар со скидкой и сам оплачивает еще и доставку (а не магазин, как в первом случае). Более понятным я считаю первое правило: если я набрал товаров на 6500 ₽, я ожидаю бесплатную доставку и скидку.

## Общий вывод

Ответьте не менее чем тремя предложениями: что вы проверили, какие границы
оказались важны и какое правило сильнее всего повлияло на итоговую сумму.

**Мой общий вывод:** В ходе выполнения домашней работы я проверила корректность работы условных операторов if/elif/else на различных сценариях расчёта стоимости заказа. Важными оказались точные граничные значения (5000 ₽ для скидки и 6000 ₽ для доставки), так как использование оператора >= вместо > напрямую определяет, получит ли покупатель скидку/бесплатную доставку. Сильнее всего на итоговую сумму повлияло правило определения порога бесплатной доставки: если рассчитывать его после применения скидки, покупатель теряет бесплатную доставку даже при крупной сумме заказа, что увеличивает итоговую сумму.


## Дополнительный уровень

Выполните **минимум два** задания. Для каждого уже подготовлены отдельные
Code- и Markdown-ячейки ниже.

1. Три уровня скидки и четыре граничных проверки.
2. Клубная карта: сравнение `and` и `or`.
3. Промокод и правило совместного применения скидок.
4. Собственное правило магазина и минимум три проверки.

In [70]:
# Расширение 1. Три уровня скидки
# До 3000 — 0%; от 3000 до 6999 — 5%; от 7000 — 10%.
# Проверьте 2999, 3000, 6999 и 7000 без циклов и списков.

order_total = 2999

if order_total >= 7000:
    discount_rate = 0.1
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Скидка: {discount_rate} Итого: {total_after_discount}")

order_total = 3000

if order_total >= 7000:
    discount_rate = 0.1
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Скидка: {discount_rate} Итого: {total_after_discount}")

order_total = 6999

if order_total >= 7000:
    discount_rate = 0.1
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Скидка: {discount_rate} Итого: {total_after_discount}")

order_total = 7000

if order_total >= 7000:
    discount_rate = 0.1
elif order_total >= 3000:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Скидка: {discount_rate} Итого: {total_after_discount}")


Сумма: 2999 Скидка: 0 Итого: 2999
Сумма: 3000 Скидка: 0.05 Итого: 2850.0
Сумма: 6999 Скидка: 0.05 Итого: 6649.049999999999
Сумма: 7000 Скидка: 0.1 Итого: 6300.0


**Вывод расширения 1:** Скидка 0% применилась при order_total = 2999, т.к. это меньше 3000 и сработала ветка else. Скидка 5% применилась при order_total = 3000 и 6999, т.к. оба значения больше или равны 3000, но меньше 7000, и сработала ветка elif. Скидка 10% применилась при order_total = 7000, т.к. это значение больше или равно 7000, и сработала ветка if. Операторы >= включают границы значения в условие.

In [71]:
# Расширение 2. Клубная карта

# Сумма достаточна и есть кубная карта
order_total = 4500
is_member = True

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000 and is_member:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"AND: Скидка {discount_rate}, Итого: {total_after_discount}")

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000 or is_member:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"OR:  Скидка {discount_rate}, Итого: {total_after_discount}")


# Сумма НЕдостаточна и есть кубная карта
order_total = 2000
is_member = True

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000 and is_member:
    discount_rate = 0.05
else:
    discount_rate = 0
total_after_discount = order_total * (1 - discount_rate)
print(f"AND: Скидка {discount_rate}, Итого: {total_after_discount}")

if order_total >= 7000:
    discount_rate = 0.10
elif order_total >= 3000 or is_member:
    discount_rate = 0.05
else:
    discount_rate = 0

total_after_discount = order_total * (1 - discount_rate)
print(f"OR:  Скидка {discount_rate}, Итого: {total_after_discount}")

# Сравните два правила: достаточная сумма AND карта; достаточная сумма OR карта.

AND: Скидка 0.05, Итого: 4275.0
OR:  Скидка 0.05, Итого: 4275.0
AND: Скидка 0, Итого: 2000
OR:  Скидка 0.05, Итого: 1900.0


**Вывод расширения 2:** Правила and и or дают одинаковый результат (скидка 5%) при order_total = 4500 и is_member = True, так как оба условия выполнены одновременно, и оба оператора возвращают True. Разные результаты появляются при order_total = 2000 и is_member = True: правило and не применило скидку (0%), т.к. условие order_total >= 3000 ложно, а False and True = False. Правило or применило скидку (5%), т.к. условие is_member истинно, а False or True = True. Таким образом, and требует выполнения обоих условий, а or - хотя бы одного

In [72]:
# Расширение 3. Промокод

# Сформулируйте правило: можно ли объединять промокод и клубную скидку?
# Присвойте discount_rate и проверьте минимум три сочетания True / False.

# и промокод и скидка
has_promo_code = True
is_member = True
if has_promo_code and is_member:
  discount_rate = 0.2
elif has_promo_code:
  discount_rate = 0.1
elif is_member:
  discount_rate = 0.15
else:
  discount_rate = 0
print(f"1. Промокод: {has_promo_code}, Карта: {is_member} Скидка: {discount_rate}")

# только скидка
has_promo_code = True
is_member = False
if has_promo_code and is_member:
  discount_rate = 0.2
elif has_promo_code:
  discount_rate = 0.1
elif is_member:
  discount_rate = 0.15
else:
  discount_rate = 0
print(f"2. Промокод: {has_promo_code}, Карта: {is_member} Скидка: {discount_rate}")

# только промокод
has_promo_code = False
is_member = True
if has_promo_code and is_member:
  discount_rate = 0.2
elif has_promo_code:
  discount_rate = 0.1
elif is_member:
  discount_rate = 0.15
else:
  discount_rate = 0
print(f"3. Промокод: {has_promo_code}, Карта: {is_member} Скидка: {discount_rate}")

# ничего
has_promo_code = False
is_member = False
if has_promo_code and is_member:
  discount_rate = 0.2
elif has_promo_code:
  discount_rate = 0.1
elif is_member:
  discount_rate = 0.15
else:
  discount_rate = 0
print(f"4. Промокод: {has_promo_code}, Карта: {is_member} Скидка: {discount_rate}")

1. Промокод: True, Карта: True Скидка: 0.2
2. Промокод: True, Карта: False Скидка: 0.1
3. Промокод: False, Карта: True Скидка: 0.15
4. Промокод: False, Карта: False Скидка: 0


**Вывод расширения 3:** Я реализовала правило повышенной скидки за комбинацию условий: при has_promo_code = True и is_member = True применяется максимальная скидка 20%. Если выполнено только одно условие (has_promo_code/is_member), применяется базовая скидка 10%/ 15% соответсмтвенно. Это приводит к тому, что клиент получает наибольшую выгоду только при одновременном использовании обоих инструментов лояльности, а магазин стимулирует покупателей собирать промокоды и вступать в клуб, так как их сочетание дает более высокую скидку, чем использование по отдельности.

In [73]:
# Расширение 4. Собственное правило магазина
# Сначала запишите правило комментарием.
# Затем реализуйте его и проверьте минимум три набора данных.

# Правило:
# Базовая скидка зависит от суммы заказа:
#   - до 2000 — 0%
#   - от 2000 до 4999 — 5%
#   - от 5000 — 10%
# Дополнительные бонусы (суммируются с базовой):
#   - наличие клубной карты: +2%
#   - промокод: +3%
# Скидка при оптовой закупке
#   - от 5 товаров: +2%
#   - от 10 товаров: +5% (вместо +2%)

# Не опт, нет карты и промокода, до 2000
order_total = 1500
is_member = False
has_promo_code = False
quantity = 2

if(order_total >= 5000):
  base_rate = 0.10
elif(order_total >= 2000):
  base_rate = 0.05
else:
  base_rate = 0

if(is_member == True):
  member_bonus = 0.02
else:
  member_bonus = 0

if(has_promo_code == True):
  promo_bonus = 0.03
else:
  promo_bonus = 0

if(quantity >= 10):
  quantity_bonus = 0.05
elif(quantity >= 5):
  quantity_bonus = 0.02
else:
  quantity_bonus = 0

discount_rate = base_rate + member_bonus + promo_bonus + quantity_bonus

if(discount_rate > 0.25):
  discount_rate = 0.25

total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Клубная карта: {is_member} Промокод: {has_promo_code} Количество: {quantity} Скидка: {discount_rate*100}% Итого: {total_after_discount}")


# опт (5), есть карта, нет промокода, заказ от 2000
order_total = 3000
is_member = True
has_promo_code = False
quantity = 5

if(order_total >= 5000):
  base_rate = 0.10
elif(order_total >= 2000):
  base_rate = 0.05
else:
  base_rate = 0

if(is_member == True):
  member_bonus = 0.02
else:
  member_bonus = 0

if(has_promo_code == True):
  promo_bonus = 0.03
else:
  promo_bonus = 0

if(quantity >= 10):
  quantity_bonus = 0.05
elif(quantity >= 5):
  quantity_bonus = 0.02
else:
  quantity_bonus = 0

discount_rate = base_rate + member_bonus + promo_bonus + quantity_bonus


total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Клубная карта: {is_member} Промокод: {has_promo_code} Количество: {quantity} Скидка: {discount_rate*100}% Итого: {total_after_discount}")


# опт (11), есть карта и промокод, заказ от 5000
order_total = 5001
is_member = True
has_promo_code = True
quantity = 11

if(order_total >= 5000):
  base_rate = 0.10
elif(order_total >= 2000):
  base_rate = 0.05
else:
  base_rate = 0

if(is_member == True):
  member_bonus = 0.02
else:
  member_bonus = 0

if(has_promo_code == True):
  promo_bonus = 0.03
else:
  promo_bonus = 0

if(quantity >= 10):
  quantity_bonus = 0.05
elif(quantity >= 5):
  quantity_bonus = 0.02
else:
  quantity_bonus = 0

discount_rate = base_rate + member_bonus + promo_bonus + quantity_bonus

if(discount_rate > 0.25):
  discount_rate = 0.25

total_after_discount = order_total * (1 - discount_rate)
print(f"Сумма: {order_total} Клубная карта: {is_member} Промокод: {has_promo_code} Количество: {quantity} Скидка: {discount_rate*100}% Итого: {total_after_discount}")


Сумма: 1500 Клубная карта: False Промокод: False Количество: 2 Скидка: 0% Итого: 1500
Сумма: 3000 Клубная карта: True Промокод: False Количество: 5 Скидка: 9.000000000000002% Итого: 2730.0
Сумма: 5001 Клубная карта: True Промокод: True Количество: 11 Скидка: 20.0% Итого: 4000.8


**Вывод расширения 4:** Я выбрала эти три набора данных, чтобы проверить работу правила на минимальном, среднем и максимальном уровнях, а также на ключевых граничных значениях. Первая проверка (order_total = 1500, без льгот, quantity = 2) доказала, что при невыполнении ни одного условия базовая ставка и все бонусы корректно остаются равными 0. Вторая проверка (order_total = 3000, is_member = True, quantity = 5) доказала, что бонусы успешно суммируются (5% + 2% + 2% = 9%) и корректно срабатывает нижняя граница оптовой скидки (ровно 5 товаров). Третья проверка (order_total = 5001, все льготы, quantity = 11) доказала, что при достижении верхних порогов  применяются максимальные тарифы (10% + 2% + 3% + 5% = 20%). Проверки доказали, что логика независимого вычисления каждого бонуса и их последующего сложения работает без ошибок

## Проверка перед сдачей

- [ ] Все обязательные задания выполнены.
- [ ] Код выполняется сверху вниз без ошибок.
- [ ] Заполнены вопрос и гипотеза.
- [ ] Завершена базовая модель заказа.
- [ ] Выполнены все четыре обязательных эксперимента.
- [ ] Проверены значения ниже, на границе и выше границы.
- [ ] В условиях изменяются значения скидки и доставки.
- [ ] Выполнены минимум два задания расширенного уровня.
- [ ] Для каждого эксперимента записан отдельный вывод.
- [ ] Сформулирован общий вывод не менее чем из трёх предложений.
- [ ] Итоговый результат выведен через f-строку.
- [ ] Файл загружен как `homework/lesson-02/solution.ipynb`.